File: src/main/java/io/Cbuild/Expansion.java
package io.Cbuild;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Expansion {

    public static class BaseExpansionEngine extends ExecIR.BaseExecIR { }

    public static class minimalApiExpansionEngine extends Expansion.BaseExpansionEngine {

        private final minimalApiValueExpansionEngine valueExpansionEngine;
        private final Env context;

        public minimalApiExpansionEngine(Env context) {
            this.valueExpansionEngine = new minimalApiValueExpansionEngine(context);
            this.context = context;
        }

        @Override
        public <T> T exec(cBuildIR.AssignmentIR ir) {
            String identifier = ir.left.exec(this.valueExpansionEngine);
            if(ir.type == cBuildIR.AssignmentType.SIMPLE) {
                String value = ir.right.exec(this.valueExpansionEngine);
                context.putRawVariable(identifier, value);
            }
            if(ir.type == cBuildIR.AssignmentType.RECURSIVE) {
                context.putDeferredVariable(identifier, ir.right);
            }

            return null;
        }
    }

    public static class minimalApiValueExpansionEngine extends Expansion.BaseExpansionEngine {

        private final Env context;

        private final Set<String> activeLookups;

        public minimalApiValueExpansionEngine(Env context) {
            this.context = context;
            this.activeLookups = new HashSet<>();
        }

        public void clearActiveLookups() {
            this.activeLookups.clear();
        }

        @Override
        @SuppressWarnings("unchecked")
        public <T> T exec(cBuildIR.ValueIR ir) {
            return (T) expandValueToString(ir, activeLookups);
        }

        @Override
        @SuppressWarnings("unchecked")
        public <T> T exec(cBuildIR.VarRefPart varRef) {
            return (T) expandVarRefToString(varRef, activeLookups);
        }

        @Override
        @SuppressWarnings("unchecked")
        public <T> T exec(cBuildIR.TextPart part) {
            return (T) part.lexeme;
        }

        private String expandValueToString(cBuildIR.ValueIR ir, Set<String> activeLookups) {
            StringBuilder builder = new StringBuilder();
            for(cBuildIR.ValuePart part : ir.parts) {
                if(part instanceof cBuildIR.VarRefPart refPart) {
                    builder.append(expandVarRefToString(refPart, activeLookups));
                }
                if(part instanceof cBuildIR.TextPart textPart) {
                    builder.append(textPart.lexeme);
                }
            }
            return builder.toString();
        }

        private String expandVarRefToString(cBuildIR.VarRefPart varRef, Set<String> activeLookups) {
            StringBuilder builder = new StringBuilder();
            for(cBuildIR.ValuePart part : varRef.nameExpr.parts) {
                if(part instanceof cBuildIR.VarRefPart refPart) {
                    builder.append(expandVarRefToString(refPart, activeLookups));
                }
                if(part instanceof cBuildIR.TextPart textPart) {
                    builder.append(textPart.lexeme);
                }
            }
            String identifier = builder.toString();

            if (activeLookups.contains(identifier)) {
                throw new cbuildException(
                        cbuildException.ErrorType.SEMANTIC,
                        "Recursive variable '" + identifier
                                + "' references itself (eventually)."
                );
            }

            if(context.symbolTable.containsKey(identifier)) {
                Env.SymbolTableVariable var = context.symbolTable.get(identifier);
                if(var.isDeferred()) {
                    cBuildIR.ValueIR valueIR =  var.getDeferredValue();
                    activeLookups.add(identifier);
                    String rawValue = exec(valueIR);
                    activeLookups.remove(identifier);
                    context.overrideVariable(identifier, Env.SymbolTableVariable.deferredVariable(valueIR, rawValue));
                    return rawValue;
                }

                return context.symbolTable.get(identifier).getRawValue();
            }
            return "";
        }
    }

    public static class minimalApiRecipeExpansionEngine extends Expansion.BaseExpansionEngine {

        private final Env context;
        public minimalApiRecipeExpansionEngine(Env context) {
            this.context = context;
        }

        @Override
        public String exec(cBuildIR.RecipeIR ir) {
            minimalApiValueExpansionEngine valueExpansionEngine
                    = new minimalApiValueExpansionEngine(this.context);
            String expandedShellCommand = valueExpansionEngine.exec( ir.command);
            return expandedShellCommand;
        }
    }

    public String expandValue(cBuildIR.ValueIR ir, Env context) {
        minimalApiValueExpansionEngine engine = new Expansion.minimalApiValueExpansionEngine(context);
        return ir.exec(engine);
    }

    public void expand(List<cBuildIR.IR> instructions, Env context) {
        minimalApiExpansionEngine expansionEngine = new minimalApiExpansionEngine(context);
        for(cBuildIR.IR ir : instructions) {
            ir.exec(expansionEngine);
        }
    }

    public void expand(cBuildIR.IR instruction, Env context) {
        minimalApiExpansionEngine expansionEngine = new minimalApiExpansionEngine(context);
        instruction.exec(expansionEngine);
    }
}

File: src/cbuild-backend/expansion.ts
import {
  AssignmentIR,
  AssignmentType,
  Executor,
  IR,
  RecipeIR,
  ValueIR,
  ValuePart,
} from "@compiler/ir.js";
import { Env, SymbolTableVariable } from "@src/cbuild-backend/env.js";

type VarRefPart = Extract<ValuePart, { kind: "variable-reference" }>;

export class RecursiveVariableExpansionException extends Error {
  public constructor(public readonly identifier: string) {
    super(`Recursive variable expansion: ${identifier}`);
    this.name = "RecursiveVariableExpansionException";
  }
}

// Adapter for the current IR executor interface; unhandled nodes return null.
export class BaseExpansionEngine implements Executor {
  public exec<T>(ir: IR | ValuePart): T {
    return null as T;
  }
}

export class ExpansionEngine extends BaseExpansionEngine {
  private readonly valueExpansionEngine: ValueExpansionEngine;
  private readonly context: Env;

  public constructor(context: Env) {
    super();
    this.valueExpansionEngine = new ValueExpansionEngine(context);
    this.context = context;
  }

  public expand<T>(ir: AssignmentIR | ValueIR | ValuePart | RecipeIR): T {
    if (ir instanceof AssignmentIR) {
      const identifier = ir.left!.exec<string>(this.valueExpansionEngine);
      if (ir.type == AssignmentType.SIMPLE) {
        const value = ir.right!.exec<string>(this.valueExpansionEngine);
        this.context.setRawVariable(identifier, value);
      }
      if (ir.type == AssignmentType.RECURSIVE) {
        this.context.setDeferredVariable(identifier, ir.right!);
      }
    }

    if (ir instanceof RecipeIR) {
      return ir.exec<T>(new RecipeExpansionEngine(this.context));
    }

    if (ir instanceof ValueIR) {
      return this.valueExpansionEngine.expand(ir) as T;
    }

    // expand value parts
    if (
      "kind" in ir &&
      (ir.kind === "variable-reference" || ir.kind === "text")
    ) {
      return this.valueExpansionEngine.expand(ir as VarRefPart) as T;
    }

    throw new Error("Unsupported IR node for expansion");
  }

  public override exec<T>(ir: IR): T {
    return this.expand(ir);
  }
}

export class ValueExpansionEngine extends BaseExpansionEngine {
  private readonly context: Env;
  private readonly activeLookups: Set<string>;

  public constructor(context: Env) {
    super();
    this.context = context;
    this.activeLookups = new Set<string>();
  }

  public clearActiveLookups(): void {
    this.activeLookups.clear();
  }

  public expand(ir: ValueIR | ValuePart): string {
    if (ir instanceof ValueIR) {
      return this.expandValueToString(ir, this.activeLookups);
    }
    if ("kind" in ir && ir.kind === "variable-reference") {
      return this.expandVarRefToString(ir, this.activeLookups);
    }
    if ("kind" in ir && ir.kind === "text") {
      return ir.lexeme;
    }

    throw new Error("Unsupported IR node or ValuePart for cbuild backend");
  }

  public override exec<T>(ir: ValueIR): T {
    return this.expandValueToString(ir, this.activeLookups) as T;
  }

  private expandValueToString(ir: ValueIR, activeLookups: Set<string>): string {
    let builder = "";
    for (const part of ir.parts) {
      if (part.kind === "variable-reference") {
        const refPart = part;
        builder += this.expandVarRefToString(refPart, activeLookups);
      }
      if (part.kind === "text") {
        const textPart = part;
        builder += textPart.lexeme;
      }
    }
    return builder.toString();
  }

  private expandVarRefToString(
    varRef: VarRefPart,
    activeLookups: Set<string>,
  ): string {
    let builder = "";
    for (const part of varRef.nameExpr.parts) {
      if (part.kind === "variable-reference") {
        const refPart = part;
        builder += this.expandVarRefToString(refPart, activeLookups);
      }
      if (part.kind === "text") {
        const textPart = part;
        builder += textPart.lexeme;
      }
    }
    const identifier = builder.toString();

    if (activeLookups.has(identifier)) {
      throw new RecursiveVariableExpansionException(identifier);
    }

    if (this.context.hasVariable(identifier)) {
      const variable = this.context.requireVariable(identifier);
      if (variable.isDeferred()) {
        const valueIR = variable.getDeferredValue()!;
        activeLookups.add(identifier);
        const rawValue = this.exec<string>(valueIR);
        activeLookups.delete(identifier);
        this.context.replaceVariable(
          identifier,
          new SymbolTableVariable(rawValue, valueIR),
        );
        return rawValue;
      }

      return this.context.requireVariable(identifier).getRawValue()!;
    }
    return "";
  }
}

export class RecipeExpansionEngine extends BaseExpansionEngine {
  private readonly context: Env;

  public constructor(context: Env) {
    super();
    this.context = context;
  }

  public override exec<T>(ir: RecipeIR): T {
    if (!(ir instanceof RecipeIR) || ir.recipe.kind !== "command") {
      throw new Error(
        "RecipeIR only support command recipes for cbuild backend",
      );
    }

    const valueExpansionEngine = new ValueExpansionEngine(this.context);
    const expandedShellCommand = valueExpansionEngine.exec<string>(
      ir.recipe.command,
    );
    return expandedShellCommand as T;
  }
}

export class Expansion {
  public expand(instructions: IR | IR[], context: Env): void {
    const expansionEngine = new ExpansionEngine(context);

    if (!Array.isArray(instructions)) {
      instructions.exec(expansionEngine);
      return;
    }

    for (const ir of instructions) {
      ir.exec(expansionEngine);
    }
  }

  public expandValue(ir: ValueIR, context: Env): string {
    const expansionEngine = new ValueExpansionEngine(context);
    return ir.exec<string>(expansionEngine);
  }
}

export function expandValue(ir: ValueIR, context: Env): string {
  const expansionEngine = new ValueExpansionEngine(context);
  return ir.exec<string>(expansionEngine);
}

export function expandRecipe(ir: RecipeIR, context: Env): string {
  const expansionEngine = new RecipeExpansionEngine(context);
  return ir.exec<string>(expansionEngine);
}
